### ISSUE TITLE
Embedd LUVI_MAIN into built applications

### ISSUE THREAD

**Issue by raphaelcohn** [luvit/luvi#80]
As I understand the code currently, the value of LUVI_MAIN is not embedded into built applications. This means that Luvi applications are sensitive to environment variables once deployed to systems outside of the builders control.
It'd be very nice that, if LUVI_MAIN is specified with LUVI_TARGET, then a simple text file containing the LUVI_MAIN value is placed in the concatenated zip. Of course, this code be in the unzipped bundle, too.
Alternatively, I'd be equally happy with disabling LUVI_MAIN for built applications, and simply relying on the symlink (main/x) with main.lua fallback.

I'm worried this could end up like LD_PRELOAD et al... a security hole waiting to be opened.

**Comment by creationix** [comment]
I'd be more concerned about `LUVI_APP` than `LUVI_MAIN`, but the general idea is correct.  There are several workflows that depend on the sensitivity to environment variables.  What about an option to make luvi apps ignore the environment variables if a certain text file exists in the bundle?  Would opt-in security be enough?

**Comment by raphaelcohn** [comment]
I'm not sure if opt in security would be enough - needs some thought. The
current design's very powerful because of the env vars, it's true, but they
are inherited by child processes - and that's the problem with the LD_* et
al stuff (likewise for other nasties, like CDPATH, and was also the cause
of the recent serious bash CVE - functions exported as env vars). Adding a
text file would be an interesting and simple solution, but I'd worry we end
up the way of java jar manifests. That was one of those 'it looked like a
good idea at the time' decisions from 15 years ago...

Whilst we're here, would it be possible to make the entry point code for
luvi do a check if it's running suid / guid, and then exit immediately?
That way no one can ever build a suid-based luvi app - a very good thing.
Nothing then stops a user using Linux capabilities, for instance, but they
have to know what they're doing to do that. Even the experts get this wrong
- qv the recent BusyBox / musl ping suid defect CVE.

Raphael Cohn
Chief Architect, stormmq
Co-Chair, OASIS MQTT Standard
Secretary, OASIS AMQP Standard
raphael.cohn@stormmq.com
+44 7590 675 756

UK Office:
Hamblethorpe Farm, Crag Lane, Bradley BD20 9DB, North Yorkshire, United
Kingdom
Telephone: +44 845 3712 567

Registered office:
16 Anchor Street, Chelmsford, Essex, CM2 0JY, United Kingdom
StormMQ Limited is Registered in England and Wales under Company Number
07175657
StormMQ.com

**Comment by creationix** [comment]
I suppose I could simply disable the environment variables if the app is bundled with a zip.

To make access to raw luvi easier, I could add a `dump` or `extract` command to lit that would split out luvi and the embedded zip into standalone files.

I'm fine with adding a setuid/ setgid check, I'm just not sure how to check for that.  PR welcome.

**Comment by creationix** [comment]
Actually, we can fix two bugs with one change here.  The environment variable wat to control luvi is annoying to begin with.  It's easy enough is bash shell terminals, but fish and windows are a bit harder.

Also if only raw luvi is going to honor these variables, they could simply be part of the arguments instead.  We can still allow developing apps like luvit that have their own arguments by shifting out the luvi arguments before calling the app's main.

For example, the following would run the luvit test suite.  The `--` argument means to pass everything after it to the luvi app itself.

```sh
luvi path/to/luvit -- path/to/luvit/tests/run.lua
```

An alias for luvit development could be done with:

```sh
alias luvit-dev='luvi path/to/luvit --'
```

At which point `luvit-dev` is a drop-in replacement command for the real `luvit`.

**Comment by creationix** [comment]
We could support `LUVI_MAIN` as a second arg to luvi itself.  I don't mind dropping `LUVI_TARGET` support since lit handles that use case better and `LUVI_TARGET` doesn't do anything that `zip` and `cat` can't do already.

**Comment by creationix** [comment]
Fixed in luvit v2.0.0 release.
